apps/src/templates/progress/progressHelpers.js

```diff
@@ -180,6 +241,176 @@ export const processedLevel = level => {
     levelNumber: level.kind === LevelKind.unplugged ? undefined : level.title,
     isConceptLevel: level.is_concept_level,
     bonus: level.bonus,
-    sublevels: level.sublevels
+    sublevels:
+      level.sublevels &&
+      level.sublevels.map(level => processedLevel(level, true))
+  };
+};
+
+/**
+ * Given a level progress object that we get from the server using either
+ * /api/user_progress or /dashboardapi/section_level_progress,
+ * extracts the result, appropriately discerning a locked/submitted
+ * result for certain levels.
+ */
+const getLevelResult = serverProgress => {
+  if (serverProgress.status === LevelStatus.locked) {
+    return TestResults.LOCKED_RESULT;
+  }
+  if (serverProgress.readonly_answers) {
+    return TestResults.READONLY_SUBMISSION_RESULT;
+  }
+  if (serverProgress.submitted) {
+    return TestResults.SUBMITTED_RESULT;
+  }
+
+  return serverProgress.result || TestResults.NO_TESTS_RUN;
+};
+
+export const levelProgressFromServer = serverObject => {
+  return {
+    status: serverObject.status || LevelStatus.not_tried,
+    result: getLevelResult(serverObject),
+    paired: serverObject.paired || false,
+    timeSpent: serverObject.time_spent || 0
   };
 };
+
+export const processServerStudentProgress = serverStudentProgress => {
+  return _.mapValues(serverStudentProgress, progress =>
+    levelProgressFromServer(progress)
+  );
+};
+
+export const processServerSectionProgress = serverSectionProgress => {
+  const studentProgress = _.mapValues(serverSectionProgress, student =>
+    processServerStudentProgress(student)
+  );
+  return studentProgress;
+};
+
+export const levelProgressWithStatus = status => {
```

Comment:
This feels like an odd extra level of indirection. Why not just do 
`levelProgressFromServer({status: activityCssClass(result)});` in line 297?

Reply:
this will make more sense in the context of part 2. there are cases (e.g. the progress legend) were we need to create `studentLevelProgressType` objects to represent a specific status, which is what this is for.

Reply:
👍